pandas/core/sparse/array.py

```diff
@@ -165,253 +206,282 @@ def __new__(cls, data, sparse_index=None, index=None, kind='integer',
+            sparse_values, sparse_index, fill_value = make_sparse(
+                data, kind=kind, fill_value=fill_value, dtype=dtype
+            )
+        else:
+            sparse_values = np.asarray(data, dtype=dtype)
+            if len(sparse_values) != sparse_index.npoints:
+                raise AssertionError("Non array-like type {type} must "
+                                     "have the same length as the index"
+                                     .format(type=type(sparse_values)))
+        # TODO: copy is unused
+
+        self._sparse_index = sparse_index
+        self._sparse_values = sparse_values
+        self._dtype = SparseDtype(sparse_values.dtype)
+        self.fill_value = fill_value
+
+    @classmethod
+    def _simple_new(cls, sparse_array, sparse_index, fill_value=None):
+        # type: (SparseArray, SparseIndex) -> 'SparseArray'
+        new = cls([])
+        new._sparse_index = sparse_index
+        new._sparse_values = sparse_array
+        new._dtype = sparse_array.dtype
 
-        if (is_integer_dtype(data) and is_float(fill_value) and
-                sp_index.ngaps > 0):
-            # if float fill_value is being included in dense repr,
-            # convert values to float
-            data = data.astype(float)
+        if fill_value is None:
+            fill_value = sparse_array.fill_value
+        new.fill_value = fill_value
+        return new
+
+    def __array__(self, dtype=None, copy=True):
+        if self.sp_index.ngaps == 0:
+            # Compat for na dtype and int values.
+            return self.sp_values
+        if dtype is None:
+            dtype = np.result_type(self.sp_values.dtype, self.fill_value)
+        out = np.full(self.shape, self.fill_value, dtype=dtype)
+        out[self.sp_index.to_int_index().indices] = self.sp_values
+        return out
 
-        result = data.view(cls)
+    def __setitem__(self, key, value):
+        # I suppose we could allow setting of non-fill_value elements.
+        msg = "SparseArray does not support item assignment via setitem"
+        raise TypeError(msg)
 
-        if not isinstance(sp_index, SparseIndex):
-            # caller must pass SparseIndex
-            raise ValueError('sp_index must be a SparseIndex')
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        return cls(scalars, dtype=dtype)
 
-        result.sp_index = sp_index
-        result._fill_value = fill_value
-        return result
+    @classmethod
+    def _from_factorized(cls, values, original):
+        return cls(values)
 
+    # ------------------------------------------------------------------------
+    # Data
+    # ------------------------------------------------------------------------
     @property
-    def _constructor(self):
-        return lambda x: SparseArray(x, fill_value=self.fill_value,
-                                     kind=self.kind)
+    def sp_index(self):
+        return self._sparse_index
```

Comment:
Would it make sense to actually only return the values of the index (the `.indices`) and not that actual object (eventually), to limit internal API exposure ?

Reply:
I don't think so, as `BlockIndex` doesn't have `.indices` it has `blocks` and `block_lengths` instead.